### PR TITLE
Implement `String#unicode_normalize` and `String#unicode_normalized?`

### LINKED ISSUE #11223: Support for Unicode normalization forms
I propose the following API for detecting and performing [Unicode normalization](https://unicode.org/reports/tr15/):

```crystal
class String
  # Returns the result of normalizing this `String` according to the given
  # [Unicode normalization form](https://unicode.org/reports/tr15/).
  def unicode_normalize(form : Unicode::NormalizationForm = :nfc) : String
  end

  # Normalizes this `String` according to the given
  # [Unicode normalization form](https://unicode.org/reports/tr15/) and
  # writes the result to the given *io*.
  def unicode_normalize(io : IO, form : Unicode::NormalizationForm = :nfc) : Nil
  end

  # Returns whether this `String` is in the given
  # [Unicode normalization form](https://unicode.org/reports/tr15/).
  def unicode_normalized?(form : Unicode::NormalizationForm = :nfc) : Bool
  end
end

enum Unicode::NormalizationForm
  # Canonical decomposition.
  NFD

  # Canonical decomposition, followed by canonical composition.
  NFC

  # Compatibility decomposition.
  NFKD

  # Compatibility decomposition, followed by canonical composition.
  NFKC
end
```

The main differences between Ruby are the `IO`-accepting overload versus their mutating overload. There are no provisions for *Stabilized Strings* (section 12) or *Stream-Safe Text Formats* (section 13).

One use case is #11222; the compiler may require all identifiers to be in a certain normalization form (most likely NFC), or automatically perform such normalization on identifiers. I did a quick search and couldn't find any existing shards that implement Unicode normalization.

### PR DESCRIPTION
**Pull request by HertzDevil**
Closes #11223.

This is a WIP; the code is in a usable state, but every non-ASCII string performs a full normalization on every call to either method. [Ways to optimize these methods](https://unicode.org/reports/tr15/#Detecting_Normalization_Forms) will be implemented soon.

The spec file directly downloads the test suite from the Unicode Character Database on each invocation. This is probably better than defining all the ~18k test cases in the spec file itself.

### CODE REVIEW

**Comment by straight-shoota** [comment]
> The spec file directly downloads the test suite from the Unicode Character Database on each invocation. This is probably better than defining all the ~18k test cases in the spec file itself.

This is inacceptable. The spec suite must be able to run without any external components.

I think we could either consider incorporating the spec source data into the repository, or make the spec to run manually.

**Comment by ysbaddaden** [comment]
That would be very nice to have! I just came across the need to remove diacritics from a String, and normalizing to NFD would make it super easy.

I'd just name the methods without the `unicode_` prefix (hence `#normalize`) but that's just a personal preference. I also wouldn't generate a String in `#normalized?` when the string is maybe normalized, but iterate each char to check if it's normalized or not, but that can be optimized later.

**Comment by HertzDevil** [comment]
I have no idea how to optimize the slow path for `#unicode_normalized?` yet. Anyone with a better algorithm in mind could take over.

**Review by straight-shoota — APPROVED** [review]

File: spec/manual/string_normalize_spec.cr [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    expected = @expected_value.join(", ") { |x| "U+%04X" % x }
+    "While testing #{@desc}:\nexpected: actual_value.codepoints != [#{expected}]"
+  end
+end
+
+# same as `assert_prints`, but uses `CodepointsEqualExpectation` instead of `eq`
+private macro assert_prints_codepoints(call, str, desc, *, file = __FILE__, line = __LINE__)
+  %str = ({{ str }}).as(String)
+  %file = {{ file }}
+  %line = {{ line }}
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/manual/string_normalize_spec.cr:34]: ping @HertzDevil

File: src/string.cr [hunk 2]
```diff
@@ -1439,6 +1439,74 @@ class String
...
+  # Normalizes this `String` according to the given
+  # [Unicode normalization form](https://unicode.org/reports/tr15/) and
+  # writes the result to the given *io*.
+  def unicode_normalize(io : IO, form : Unicode::NormalizationForm = :nfc) : Nil
+    return io << self if Unicode.quick_check_normalized(self, form).yes?
+    do_unicode_normalize(io, form)
+  end
+
+  private def do_unicode_normalize(io, form)
+    codepoints = [] of Int32
```
[the remark this answers is not in the document]
Comment (HertzDevil) [reply src/string.cr:1466]: The specification does say there is an upper bound of `3 * size` codepoints for any normalization form, but that would be overkill for the majority of strings.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**